### PR TITLE
Remove unneeded locking

### PR DESCRIPTION
**Pull request by arkpar**
Reduced lock contention a bit.
There is no global lock on the blockchain anymore, instead there is a lock on the `insert_block` function that guarantees that all updates are atomic.
Reads are are also atomic: database changes are committed with the transaction. Most of the caches that are addressed by block hash can be updated individually. The only exceptions are `best_block`, `block_hashes` and `transaction_addresses`. These are updated alltogether.
Block queue has been optimized to use more fine grained locking.
Global lock on the block queue has been removed. It was not needed in the first place since the module is fully reenterable.

### CODE REVIEW

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -753,8 +770,16 @@ impl BlockChain {
 
 				// TODO: handle block_hashes properly.
 				block_hashes.clear();
+
+				blocks.shrink_to_fit();
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: why?

Comment (arkpar) [inline ethcore/src/blockchain/blockchain.rs]: Removing something from a `HashMap` does not guarantee the memory will be deallocated. We use `heapsize` to measure cache size on line 782. The condition there would always be false after max size is reached for the first time.

File: ethcore/src/block_queue.rs [hunk 2]
```diff
@@ -278,27 +288,29 @@ impl BlockQueue {
...
+	pub fn import_block(&self, bytes: Bytes) -> ImportResult {
 		let header = BlockView::new(&bytes).header();
 		let h = header.hash();
-		if self.processing.read().unwrap().contains(&h) {
-			return Err(ImportError::AlreadyQueued);
-		}
 		{
-			let mut verification = self.verification.lock().unwrap();
-			if verification.bad.contains(&h) {
+			if self.processing.read().unwrap().contains(&h) {
```
Comment (NikVolf) [inline ethcore/src/block_queue.rs]: Any rational explanation of introducing this block around `if` expression?

File: util/sha3/build.rs [hunk 3]
```diff
@@ -21,6 +21,6 @@
 extern crate gcc;
 
 fn main() {
-    gcc::compile_library("libtinykeccak.a", &["src/tinykeccak.c"]);
+	gcc::Config::new().file("src/tinykeccak.c").flag("-O3").compile("libtinykeccak.a");
```
Comment (NikVolf) [inline util/sha3/build.rs]: not related to the pr subject

**Pull request merged**